File: 백준/Gold/6198. 옥상 정원 꾸미기/옥상 정원 꾸미기.java
import java.util.*;
import java.io.*;
import java.math.*;

public class Main {

    static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

    static int N;
    static int[] hs;
    static Stack<Integer> stack = new Stack<>();

    static void init() throws Exception {
        N = Integer.parseInt(reader.readLine()); // 빌딩 수. 1 ~ 8만
        hs = new int[N+1];

        // 빌딩 높이 입력. 1 ~ 10억
        for(int i=1; i<=N; i++) hs[i] = Integer.parseInt(reader.readLine());
    }

    static void calc() throws Exception {
        long result = 0;
        for (int h: hs) {
            while (!stack.isEmpty() && stack.peek() <= h)
                stack.pop();

            result += stack.size();

            stack.push(h);
        }

        System.out.println(result);
    }

    public static void main(String[] args) throws Exception {
        init();
        calc();
    }

}

File: 백준/Gold/6198. 옥상 정원 꾸미기/옥상 정원 꾸미기.cc
#include <iostream>
#include <string>
#include <vector>
using namespace std;

void init(int& N, vector<int>& hs) {
    cin >> N;
    
    int tmp;
    for(int i=0; i<N; i++) {
        cin >> tmp;
        hs.push_back(tmp);
    }

}

long calc(vector<int>& hs) {

    long result = 0;

    vector<int> stack;

    for(int h: hs) {
        
        auto it = stack.rbegin();
        while (it != stack.rend() && (*it <= h)) {
            stack.pop_back();
            it++;
        }

        result += stack.size();

        stack.push_back(h);
    }

    return result;
}

int main() {

    int N;
    vector<int> hs;
    init(N, hs);

    long result = calc(hs);
    cout << result << endl;

    return 0;
}
